File: app/Models/Grade.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Grade extends Model
{
    protected $fillable = [
        'name', 'name_ar', 'level_order', 'description'
    ];

    public function sections()
    {
        return $this->hasMany(Section::class);
    }

    public function subjects()
    {
        return $this->belongsToMany(Subject::class, 'grade_subject');
    }

    public function stages()
    {
        return $this->belongsToMany(Stage::class, 'grade_stage');
    }

    public function subjectsForReligion(?string $religion)
    {
        $subjects = $this->subjects()->with('children')->get();

        return $subjects->flatMap(function ($subject) use ($religion) {
            if ($subject->is_religion_based && $subject->children->count()) {
                $child = $subject->children->where('religion', $religion)->first();
                return $child ? [$child] : [];
            }
            if ($subject->parent_id && $subject->parent && $subject->parent->is_religion_based) {
                return [];
            }
            return [$subject];
        });
    }
}

File: flutter/lib/models/grade.dart
class Grade {
  final int id;
  final String name;
  final String nameAr;
  final int? levelOrder;
  final String? description;

  Grade({
    required this.id,
    required this.name,
    required this.nameAr,
    this.levelOrder,
    this.description,
  });

  factory Grade.fromJson(Map<String, dynamic> json) => Grade(
    id: json['id'],
    name: json['name'] ?? '',
    nameAr: json['name_ar'] ?? '',
    levelOrder: json['level_order'],
    description: json['description'],
  );

  Map<String, dynamic> toJson() => {
    'name': name,
    'name_ar': nameAr,
    'level_order': levelOrder,
    'description': description,
  };
}
